docs/import/manage_disaster_flexible.js

```diff
@@ -1,22 +1,154 @@
 
 export {
-  finishPending,
   initializeFlexibleDisaster,
+  PendingChecker,
   setUpFlexibleOnPageLoad,
-  startPending,
   useDamageForBuildings,
   validateFlexibleUserFields,
 };
+// For testing.
+export {componentsData, POVERTY_BUILDINGS_PATH};
+
+const PendingState = Object.freeze({
+  NOT_PENDING: 0,
+  PENDING: 1,
+  PENDING_BUT_INVISIBLE: 2,
+});
+
+/**
+ * Utility to note start/end of asynchronous operations that should delay
+ * validation. Needed because if user switches the disaster or asset value,
+ * currently pending operations may no longer be needed. For instance, if we are
+ * waiting for a disaster listing and the user changes the disaster, we'll
+ * immediately start waiting for the new disaster's listing. The old disaster
+ * listing status is irrelevant. When the first disaster was initialized, we
+ * called {@link maybeStartPending}, which triggered a "pending" status. When
+ * the new disaster initializes, we call it again. Because the first call had
+ * not finished, we don't increment the pending count. Whenever the first call
+ * finishes, it will not call {@link countDownPendingOperations}: only the
+ * current disaster is allowed to do that.
+ *
+ * Similarly, if an asset value changes, we'll immediately kick off a query for
+ * the new values' columns. If the old value column query had not completed, we
+ * don't increment the pending count.
+ *
+ * There should be one `PendingChecker` for each operation that must finish
+ * before validation can run, but which can be rendered irrelevant by a user
+ * action. Thus, each select whose value cascades (poverty, geography, damage,
+ * buildings) must have a `PendingChecker`. Similarly initialization as a whole
+ * has a `PendingChecker`. This keeps our count of pending operations limited to
+ * the types of operations there can be, which is correct: the user can never be
+ * waiting on more than one operation of a given type.
+ *
+ * A subtle issue arises when a pending value becomes unnecessary: for instance,
+ * if the div that needs to be populated is hidden. Then we immediately count
+ * down the pending operations via {@link markInvisible}. But if the div ever
+ * becomes visible again, and the operation has still not finished, we count
+ * pending operations back up using {@link markVisible}. This allows listings to
+ * complete in the background and be ready, even if the user is switching other
+ * options back and forth.
+ */
+class PendingChecker {
+  pending = PendingState.NOT_PENDING;
+
+  /** @return {boolean} False if this checker was already pending */
+  maybeStartPending() {
+    if (!isFlexible()) {
+      return true;
```

Comment:
Why early return if state based?

Reply:
Because this whole file is only for flexible disasters